authorize_error.go

```diff
@@ -34,47 +34,49 @@ func (f *Fosite) WriteAuthorizeError(rw http.ResponseWriter, ar AuthorizeRequest
 	rw.Header().Set("Pragma", "no-cache")
 
 	rfcerr := ErrorToRFC6749Error(err)
+	if !f.SendDebugMessagesToClients {
+		rfcerr = rfcerr.Sanitize()
+	}
+
 	if !ar.IsRedirectURIValid() {
-		if !f.SendDebugMessagesToClients {
-			rfcerr.Debug = ""
-		}
+		rw.Header().Set("Content-Type", "application/json;charset=UTF-8")
 
 		js, err := json.MarshalIndent(rfcerr, "", "\t")
 		if err != nil {
-			http.Error(rw, err.Error(), http.StatusInternalServerError)
+			if f.SendDebugMessagesToClients {
+				http.Error(rw, fmt.Sprintf(`{\n\t"error": "server_error",\n\t"error_description": "%s"\n}`, err.Error()), http.StatusInternalServerError)
+			} else {
+				http.Error(rw, `{\n\t"error": "server_error"\n}`, http.StatusInternalServerError)
+			}
 			return
 		}
 
-		rw.Header().Set("Content-Type", "application/json;charset=UTF-8")
 		rw.WriteHeader(rfcerr.Code)
 		rw.Write(js)
 		return
 	}
 
 	redirectURI := ar.GetRedirectURI()
-	query := url.Values{}
-	query.Add("error", rfcerr.Name)
-	query.Add("error_description", rfcerr.Description)
-	query.Add("state", ar.GetState())
-	if f.SendDebugMessagesToClients && rfcerr.Debug != "" {
-		query.Add("error_debug", rfcerr.Debug)
-	}
 
-	if rfcerr.Hint != "" {
-		query.Add("error_hint", rfcerr.Hint)
-	}
+	// The endpoint URI MUST NOT include a fragment component.
+	redirectURI.Fragment = ""
+
+	query := rfcerr.ToValues()
+	query.Add("state", ar.GetState())
 
+	var redirectURIString string
 	if !(len(ar.GetResponseTypes()) == 0 || ar.GetResponseTypes().ExactOne("code")) && errors.Cause(err) != ErrUnsupportedResponseType {
-		redirectURI.Fragment = query.Encode()
+		redirectURIString = redirectURI.String() + "#" + query.Encode()
```

Comment:
We have to manually URL encode the fragment (by the spec) and append it. Previously, we double escaped stuff here. (There is no `RawFragment` before go 1.15, and `RawFragment` in go 1.15 is different in behavior from `RawQuery`.)